infra/terraform/global/assets/create_etcd_ebs_snapshot/main.go

```diff
@@ -0,0 +1,82 @@
+package main
+
+import (
+	"./snapshot"
+	"./tag"
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"log"
+	"os"
+)
+
+var (
+	clientSession = session.Must(session.NewSession())
+	connection    = ec2.New(clientSession)
+)
+
+func main() {
+
+	// Filter instances with role set to master
+	instanceKeyFilter := os.Getenv("INSTANCE_TAG_KEY")
+	instanceValueFilter := os.Getenv("INSTANCE_TAG_VALUE")
+
+	instanceInput := &ec2.DescribeInstancesInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String(fmt.Sprintf("tag:%s", instanceKeyFilter)),
+				Values: []*string{aws.String(instanceValueFilter)},
+			},
+		},
+	}
+
+	// A reservation corresponds to a command to start instances
+	// A reservation is what you do to provision instances, while an instance is what you get
+	results, err := connection.DescribeInstances(instanceInput)
+	if err != nil {
+		log.Fatal(err.Error())
+	}
+
+	for _, res := range results.Reservations {
+
+		for _, instance := range res.Instances {
+
+			for _, blk := range instance.BlockDeviceMappings {
+
+				// Exclude root volumes.  Root volumes have "Delete on termination" set to true mostly
+				isRootDevice := *blk.Ebs.DeleteOnTermination
+
+				if isRootDevice != true {
+					vol_id := blk.Ebs.VolumeId
+
+					// Perform and tag the snapshot
+					snapShot := create_volume_snapshot.CreateSnapshot(connection, vol_id)
+
+					getVolumesTags := tag.FetchResourceTags(connection, vol_id)
+
+					snapShotID := *snapShot.SnapshotId
+					instanceName := *instance.Tags[1].Value
```

Comment:
This assumption here is that `instance.Tags[1]` is the `Instance_Name` one.

Is this reliable? e.g. don't know if relying on order is sound.
I had a quick look at the [`ec2.DescribeTagsOutput` type documentation](https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#DescribeTagsOutput) and it doesn't look like it has something like a "GetTag" method, maybe it's the case of loop through the `Tags` ourselves which yes, it's a bit annoying.

Reply:
Probably similar to scenario in this SO question: https://stackoverflow.com/questions/38654383/how-to-search-for-an-element-in-a-golang-slice

Reply:
Essentially something like this (not tested):

```go
					var instanceName string
					for _, tag := getVolumesTags.Tags {
						if tag.Key == "Instance_Name" {
							instanceName = tag.Value
							break
						}
					}
					if instanceName == "" {
						// log error "Instance doesn't have 'Instance_Name' tag. Can't tag snapshot"
						continue
					}
```

Reply:
Funny enough, I checked our snapshots and you're right.  the instance name is not always returned.  Only sometimes. I'll put your suggestion in. Thanks